test/CardPaymentProcessorUpgradeable.test.ts

```diff
@@ -0,0 +1,1461 @@
+    }
+  }
+
+  async function setExecutorRole(account: SignerWithAddress) {
+    await proveTx(cardPaymentProcessor.grantRole(executorRole, account.address));
+  }
+
+  async function clearPayments(payments: TestPayment[], adminAccount: SignerWithAddress) {
+    const authorizationIds: string[] = [];
+    payments.forEach((payment: TestPayment) => {
+      authorizationIds.push(createBytesString(payment.authorizationId, BYTES16_LENGTH));
+      payment.status = PaymentStatus.Cleared;
+    });
+    await proveTx(cardPaymentProcessor.connect(adminAccount).clearPayments(authorizationIds));
+  }
+
+  function defineBalancesPerAccount(payments: TestPayment[], targetPaymentStatus: PaymentStatus): Map<string, number> {
+    const balancesPerAccount: Map<string, number> = new Map<string, number>();
+
+    payments.forEach((payment: TestPayment) => {
+      const address: string = payment.account.address;
+      let newBalance: number = balancesPerAccount.get(address) || 0;
+      if (payment.status == targetPaymentStatus) {
+        newBalance += payment.amount;
+      }
+      balancesPerAccount.set(address, newBalance);
+    });
+
+    return balancesPerAccount;
+  }
+
+  function defineExpectedCardPaymentProcessorState(payments: TestPayment[]): CardPaymentProcessorState {
+    const tokenBalance: number = countNumberArrayTotal(
+      payments.map(
+        function (payment: TestPayment): number {
+          return payment.status == PaymentStatus.Uncleared || payment.status == PaymentStatus.Cleared
+            ? payment.amount
+            : 0;
+        }
+      )
+    );
+    const totalClearedBalance: number = countNumberArrayTotal(
+      payments.map(
+        function (payment: TestPayment): number {
+          return payment.status == PaymentStatus.Cleared ? payment.amount : 0;
+        }
+      )
+    );
+    const totalUnclearedBalance: number = countNumberArrayTotal(payments.map(
+        function (payment: TestPayment): number {
+          return payment.status == PaymentStatus.Uncleared ? payment.amount : 0;
+        }
+      )
+    );
+    const clearedBalancesPerAccount: Map<string, number> =
+      defineBalancesPerAccount(payments, PaymentStatus.Cleared);
+    const unclearedBalancesPerAccount: Map<string, number> =
+      defineBalancesPerAccount(payments, PaymentStatus.Uncleared);
+
+    return {
+      tokenBalance,
+      totalUnclearedBalance,
+      totalClearedBalance,
+      clearedBalancesPerAccount,
+      unclearedBalancesPerAccount,
+    };
+  }
+
+  async function checkPaymentStructuresOnBlockchain(payments: TestPayment[]) {
+    for (let i = 0; i < payments.length; ++i) {
+      const expectedPayment: TestPayment = payments[i];
+      const actualPayment = await cardPaymentProcessor.paymentFor(
+        createBytesString(expectedPayment.authorizationId, BYTES16_LENGTH)
+      );
+      checkEquality(actualPayment, expectedPayment, i);
+      if (!!expectedPayment.parentTxHash) {
```

Comment:
Codacy has a fix for the issue: [Redundant double negation.](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
      if (expectedPayment.parentTxHash) {
```